// griddle-react/modules/gridRow.jsx.js
/*
   See License / Disclaimer https://raw.githubusercontent.com/DynamicTyped/Griddle/master/LICENSE
*/
var React = require('react');
var _ = require('underscore');

var GridRow = React.createClass({displayName: "GridRow",
    getDefaultProps: function(){
      return {
        "isChildRow": false,
        "showChildren": false,
        "data": {},
        "columns" : [],
        "metadataColumns": [],
        "hasChildren": false,
        "columnMetadata": null,
        "useGriddleStyles": true,
        "useGriddleIcons": true,
        "isSubGriddle": false,
        "parentRowCollapsedClassName": "parent-row",
        "parentRowExpandedClassName": "parent-row expanded",
        "parentRowCollapsedComponent": "▶",
        "parentRowExpandedComponent": "▼"
      }
    },
    handleClick: function(){
      this.props.toggleChildren();
    },
    render: function() {
        var that = this;
        var columnStyles = this.props.useGriddleStyles ?
          {
            padding: "5px",
            backgroundColor: "#FFF",
            borderTopColor: "#DDD",
            color: "#222"
          } : null;

        var data = _.pairs(this.props.columns.length === 0 ? this.props.data : _.pick(this.props.data, this.props.columns))

        var nodes = data.map(function(col, index) {
            var returnValue = null;
            var meta = _.findWhere(that.props.columnMetadata, {columnName: col[0]});

            //todo: Make this not as ridiculous looking
            firstColAppend = index === 0 && that.props.hasChildren && that.props.showChildren === false && that.props.useGriddleIcons ?
              React.createElement("span", {style: that.props.useGriddleStyles&&{fontSize: "10px", marginRight:"5px"}}, that.props.parentRowCollapsedComponent) :
              index === 0 && that.props.hasChildren && that.props.showChildren && that.props.useGriddleIcons ?
                React.createElement("span", {style: that.props.useGriddleStyles&&{fontSize: "10px"}}, that.props.parentRowExpandedComponent) : "";

            if(index === 0 && that.props.isChildRow && that.props.useGriddleStyles){
              columnStyles = _.extend(columnStyles, {paddingLeft:10})
            }


            if (that.props.columnMetadata !== null && that.props.columnMetadata.length > 0 && typeof meta !== "undefined"){
              var colData = (typeof meta === 'undefined' || typeof meta.customComponent === 'undefined' || meta.customComponent === null) ? col[1] : React.createElement(meta.customComponent, {data: col[1], rowData: that.props.data});
              returnValue = (meta == null ? returnValue : React.createElement("td", {onClick: that.props.hasChildren && that.handleClick, className: meta.cssClassName, key: index, style: columnStyles}, colData));
            }

            return returnValue || (React.createElement("td", {onClick: that.props.hasChildren && that.handleClick, key: index, style: columnStyles}, firstColAppend, col[1]));
        });

        //this is kind of hokey - make it better
        var className = "standard-row";


        if(that.props.isChildRow){
            className = "child-row";
        } else if (that.props.hasChildren){
            className = that.props.showChildren ? this.props.parentRowExpandedClassName : this.props.parentRowCollapsedClassName;
        }

        return (React.createElement("tr", {className: className}, nodes));
    }
});

module.exports = GridRow;
